file_store.go

```diff
@@ -0,0 +1,258 @@
+	}
+	return authCfg.Credential()
+}
+
+// Put saves credentials into the store for the given server address.
+// Returns ErrPlaintextSaveDisabled if fs.DisableSave is set to true.
+func (fs *FileStore) Put(_ context.Context, serverAddress string, cred auth.Credential) error {
+	if fs.DisableSave {
+		return ErrPlaintextSaveDisabled
+	}
+
+	fs.rwLock.Lock()
+	defer fs.rwLock.Unlock()
+
+	authCfg := newAuthConfig(cred)
+	authCfgBytes, err := json.Marshal(authCfg)
+	if err != nil {
+		return fmt.Errorf("failed to marshal auth field: %w", err)
+	}
+	fs.authsCache[serverAddress] = authCfgBytes
+	return fs.saveFile()
+}
+
+// Delete removes credentials from the store for the given server address.
+func (fs *FileStore) Delete(ctx context.Context, serverAddress string) error {
+	fs.rwLock.Lock()
+	defer fs.rwLock.Unlock()
+
+	if _, ok := fs.authsCache[serverAddress]; !ok {
+		// no ops
+		return nil
+	}
+	delete(fs.authsCache, serverAddress)
+	return fs.saveFile()
+}
+
+// saveFile saves fs.content into fs.configPath.
+func (fs *FileStore) saveFile() (returnErr error) {
+	// marshal content
+	authsBytes, err := json.Marshal(fs.authsCache)
+	if err != nil {
+		return fmt.Errorf("failed to marshal credentials: %w", err)
+	}
+	fs.content[configFieldAuths] = authsBytes
+	jsonBytes, err := json.MarshalIndent(fs.content, "", "\t")
+	if err != nil {
+		return fmt.Errorf("failed to marshal config: %w", err)
+	}
+
+	// write the content to a ingest file for atomicity
+	configDir := filepath.Dir(fs.configPath)
+	if err := os.MkdirAll(configDir, 0700); err != nil {
+		return fmt.Errorf("failed to make directory %s: %w", configDir, err)
+	}
+	ingest, err := ioutil.Ingest(configDir, bytes.NewReader(jsonBytes))
+	if err != nil {
+		return fmt.Errorf("failed to save config file: %w", err)
+	}
+	defer func() {
+		if returnErr != nil {
+			// clean up the ingest file in case of error
+			os.Remove(ingest)
+		}
+	}()
+
+	// handle symlink
+	targetPath := fs.configPath
+	if link, err := os.Readlink(fs.configPath); err == nil {
+		targetPath = link
+	}
+	// copy file with original ownership and permissions
+	if err := ioutil.CopyFilePermissions(targetPath, ingest); err != nil {
+		return fmt.Errorf("failed to copy permissions from %s to %s: %w", targetPath, ingest, err)
+	}
+	// overwrite the target file
+	if err := os.Rename(ingest, targetPath); err != nil {
+		return fmt.Errorf("failed to save config file: %w", err)
+	}
+	return nil
+}
+
+// encodeAuth base64-encodes username and password into base64(username:password).
+func encodeAuth(username, password string) string {
+	if username == "" && password == "" {
+		return ""
+	}
+	return base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+}
+
+// decodeAuth decodes a base64 encoded string and returns username and password.
+func decodeAuth(authStr string) (username string, password string, err error) {
+	if authStr == "" {
+		return "", "", nil
+	}
```

Comment:
This can be covered by `DecosdeString()`, can't it?

Reply:
We still need it for early return.